text/3263-pre-release-sticky.md

```diff
@@ -0,0 +1,115 @@
+
+Finally, contrary to release version where most user benefits from having cargo pick the latest compatible version as default, there is little sense to do the same for pre-release version cause there are very likely be breaking change between all pre-release version. This is a trap that most user don't expect to happen (why does my project don't compile anymore ?!?). Pre-release version precedence of `Cargo.toml` should default to something more user friendly. User expect bugs from a pre-release but no one want their build suddenly break when a new pre-release is out.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+By default Cargo follow Semver precedence rules for release of a crate but will make no assumption for pre-release.
+
+```toml
+[dependencies.foo]
+version = "1.0.0-alpha.0"
+```
+
+will now be considered by Cargo as:
+
+```toml
+[dependencies.foo]
+version = "=1.0.0-alpha.0"
+```
+
+while release version like:
+
+```toml
+[dependencies.foo]
+version = "1.0.0"
+```
+
+are considered by Cargo as:
+
+```toml
+[dependencies.foo]
+version = "^1.0.0"
+```
+
+The latter doesn't change.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Do it transparently
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+ * This could break workflow
+ * It's make pre-release semver rules have a different behavior than release for Cargo.
+ * It's a change that affect the whole Rust ecosystem.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+In practice, do it transparently should not break any heathy workflow, user that put a pre-release version clearly want to use this version and if user have been using a most recent pre-release version not on purpose it's very unlikely their build would break. We could check the impact of implementing this transparently with a [crater](https://github.com/rust-lang/crater) run.
+
+To introduce this change we could also:
+
+  1. Use the `resolver` value version, bump it to `3`:
+   
+     This may be overkill but it's a naturel way to make change in the way semver rule are trated by Cargo.
+     
+  2. Use a separate value `pre-release-updates = "sticky" # or "default"`:
+
+     This is less overkill, but have the drawback to include a new value in `Cargo.toml` just for that.
+
+Both these solutions would need a new Rust edition to be introduce by default.
+
+Instead, We could change the rule of Semver for pre-release and say there is no precedence rule for pre-release. `1.0.0-alpha.0` would never match any other requirement that exact same version. That mean that `^1.0.0-alpha.0` could only match `1.0.0-alpha.0` version. This have the major benefit to not introduce inconsistency with pre-release and release in Cargo resolve but this would not follow Semver. This could also be adopted transparently or using opt-in solution.
```

Comment:
It's not clear to me why this "would not follow SemVer". Also I don't think this has to do with precedence. As I read the SemVer spec, the only thing it says is that incompatible releases should always bump the major version. This is already not how Cargo works (since it considers 0.1.0 and 0.2.0 incompatible), so I'm not sure it's a strong argument.

Reply:
I see the confusion now, I read it as "what do when user want `^1.0.0` that what have been implemented in Cargo so far. I will try to word it as "compatible version".